core/src/ancestor_hashes_service.rs

```diff
@@ -0,0 +1,364 @@
+        Self { thread_hdls }
+    }
+
+    pub fn join(self) -> thread::Result<()> {
+        for thread_hdl in self.thread_hdls {
+            thread_hdl.join()?;
+        }
+        Ok(())
+    }
+
+    /// Listen for responses to our ancestors hashes repair requests
+    fn run_responses_listener(
+        ancestor_hashes_request_statuses: Arc<DashMap<Slot, DeadSlotAncestorRequestStatus>>,
+        response_receiver: PacketReceiver,
+        blockstore: Arc<Blockstore>,
+        outstanding_requests: Arc<RwLock<OutstandingAncestorHashesRepairs>>,
+        exit: Arc<AtomicBool>,
+        duplicate_slots_reset_sender: DuplicateSlotsResetSender,
+    ) -> JoinHandle<()> {
+        Builder::new()
+            .name("solana-ancestor-hashes-responses-service".to_string())
+            .spawn(move || {
+                let mut last_stats_report = Instant::now();
+                let mut stats = AncestorHashesResponsesStats::default();
+                let mut max_packets = 1024;
+                loop {
+                    let result = Self::process_new_responses(
+                        &ancestor_hashes_request_statuses,
+                        &response_receiver,
+                        &blockstore,
+                        &outstanding_requests,
+                        &mut stats,
+                        &mut max_packets,
+                        &duplicate_slots_reset_sender,
+                    );
+                    match result {
+                        Err(Error::RecvTimeout(_)) | Ok(_) => {}
+                        Err(err) => info!("ancestors hashes reponses listener error: {:?}", err),
+                    };
+                    if exit.load(Ordering::Relaxed) {
+                        return;
+                    }
+                    if last_stats_report.elapsed().as_secs() > 2 {
+                        stats.report();
+                        last_stats_report = Instant::now();
+                    }
+                }
+            })
+            .unwrap()
+    }
+
+    /// Process messages from the network
+    fn process_new_responses(
+        ancestor_hashes_request_statuses: &DashMap<Slot, DeadSlotAncestorRequestStatus>,
+        response_receiver: &PacketReceiver,
+        blockstore: &Blockstore,
+        outstanding_requests: &RwLock<OutstandingAncestorHashesRepairs>,
+        stats: &mut AncestorHashesResponsesStats,
+        max_packets: &mut usize,
+        duplicate_slots_reset_sender: &DuplicateSlotsResetSender,
+    ) -> Result<()> {
+        let timeout = Duration::new(1, 0);
+        let mut responses = vec![response_receiver.recv_timeout(timeout)?];
+        let mut total_packets = responses[0].packets.len();
+
+        let mut dropped_packets = 0;
+        while let Ok(more) = response_receiver.try_recv() {
+            total_packets += more.packets.len();
+            if total_packets < *max_packets {
+                // Drop the rest in the channel in case of DOS
```

Comment:
Probably unlikely but if for whatever reason we have 50 ancestor slot repairs and we receive all 21 samples for each at the same time what would happen here? Do we have any case to ask for the missing samples again?

Reply:
hmmmm I think the ancestor request in the pending queue would time out and and we would try again